kedro/contrib/io/matplotlib/matplotlib_s3_writer.py

```diff
@@ -0,0 +1,162 @@
+from typing import Any, Dict, List, Optional, Union
+
+from matplotlib.pyplot import figure
+from s3fs import S3FileSystem
+
+from kedro.io import AbstractDataSet, DataSetError
+
+
+class MatplotlibWriterS3(AbstractDataSet):
+    # pylint: disable=too-many-instance-attributes
+    """
+    ``MatplotlibWriter`` saves matplotlib objects as image files.
+
+    Example:
+    ::
+
+        import matplotlib.pyplot as plt
+        from kedro.contrib.io.matplotlib_s3_writer import MatplotlibWriterS3
+
+        # Saving single plot
+        plt.plot([1, 2, 3], [4, 5, 6])
+        single_plot_writer = MatplotlibWriterS3(
+            bucket="my-super-great-bucket", filepath="matplot_lib_single_plot.png"
+        )
+        single_plot_writer.save(plt)
+
+        # Saving dictionary of plots (with SSE)
+        plots_dict = {}
+        for colour in ["blue", "green", "red"]:
+            plots_dict[colour] = plt.figure()
+            plt.plot([1, 2, 3], [4, 5, 6], color=colour)
+            plt.close()
+        dict_plot_writer = MatplotlibWriterS3(
+            bucket="my-super-great-bucket", filepath="matplotlib_dict",
+        )
+        dict_plot_writer.save(plots_dict)
+
+        # Saving list of plots
+        plots_list = []
+        for index in range(5):
+            plots_list.append(plt.figure())
+            plt.plot([1,2,3],[4,5,6], color=colour)
+        list_plot_writer = MatplotlibWriterS3(
+            bucket="my-super-great-bucket", filepath="matplotlib_list"
+        )
+        list_plot_writer.save(plots_list)
+    """
+
+    # pylint: disable=too-many-arguments
+    def __init__(
+        self,
+        bucket_name: str,
+        filepath: str,
+        s3fs_args: Optional[Dict] = None,
+        credentials: Optional[Dict[str, Any]] = None,
+        savefig_args: Dict[str, Any] = None,
+    ) -> None:
+        """Creates a new instance of ``MatplotlibWriter``.
+
+        Args:
+            bucket_name: Name of the bucket without "s3://" prefix
+            filepath: Path to a matplot object file.
+            s3fs_args: Dictionary of arguments
+            credentials: A dictionary of s3 access and secret keys.
+                Must contain ``aws_access_key_id`` and ``aws_secret_access_key``.
+                Updates ``s3_client_args`` if provided.
+            savefig_args: Save args passed to `plt.savefig`. See
+                https://matplotlib.org/api/_as_gen/matplotlib.pyplot.savefig.html
+        """
+
+        _credentials = copy.deepcopy(credentials) or {}
+
+        self._s3fs_args = s3fs_args or {}
+
+        if _credentials:
+            if "client_kwargs" not in self._s3fs_args.keys():
+                self._s3fs_args["client_kwargs"] = {}
+
+            self._s3fs_args["client_kwargs"]["aws_access_key_id"] = _credentials[
+                "aws_access_key_id"
+            ]
+            self._s3fs_args["client_kwargs"]["aws_secret_access_key"] = _credentials[
+                "aws_secret_access_key"
+            ]
```

Comment:
Credentials should not be provided in `catalog.yml`, via `s3fs_args`, so I would strongly encourage to keep the 2 options separate. I'd remove this bit of code completely, leaving AWS keys delegated to `credentials`, and other optional arguments to `s3fs_args`.

Reply:
This is related to the change suggested below as well, putting here for easier reference.
`_s3 = S3FileSystem(client_kwargs=_credentials, **self._s3fs_args)`

I'm a bit confused. `credentials` are provided separately:
```
    def __init__(
        self,
        bucket_name: str,
        filepath: str,
        s3fs_args: Optional[Dict] = None,
        credentials: Optional[Dict[str, Any]] = None,
        savefig_args: Dict[str, Any] = None,
    ) -> None:
```
I just added them to the `client_kwargs` so that I do not lose the ability to provide other options to `client_kwargs`.

Reply:
I get you - would these extra options qualify to be passed in through credentials directly, like the aws keys? Looking at the API of `Session.client` in `boto3`, where the `client_kwargs` are passed, that would include: region_name, api_version, verify, endpoint_url, aws_session_token. The last one for instance should definitely be treated as credential.

Reply:
Hmmm... Yes, I see..

So, should it be:
```
    def __init__(
        self,
        ...
        # docstring tries to point out that for client_kwargs should be fed in here
        credentials: Optional[Dict[str, Any]] = None, 
        ...
    ) -> None:
```

Or
```
So, should it be:
    def __init__(
        self,
        ...
        # docstring tries to point out that keys should be feed in here via credentials
        client_kwargs: Optional[Dict[str, Any]] = None,
        ...
    ) -> None:
```

1st one is more consistent with the rest of the classes. But option 2 I think is a bit clearer and matches the underlying library a bit better.

Reply:
The way I see it, the user/interface doesn't need to know about the underlying backend, and the second approach would make the two (interface and backend) too tightly coupled. If we one day decide, for whatever reason, that we don't use `s3fs` but something else new and cool, `client_kwargs` won't make sense anymore and we're left with either tech debt or breaking changes. So I'd prefer the first approach, maybe specifying in docs what you can put in there/where it goes, you think it's useful. It also means it stays consistent with the other datasets, where the way to specify options for the client would still be via `credentials`.

Reply:
Yeah makes sense! Done